File: ios/CruiseMesh/Mesh/InboundEngineSettings.swift
import Foundation

/// Which engine dispositions one arriving envelope.
///
/// One selection per envelope, and deliberately no third value. The inbound
/// path's unit of work is a single frame — gate, open, deliver-or-carry,
/// re-flood, record — and a frame whose gate came from one engine and whose
/// carry came from the other has no single owner for the `seen` record it left
/// behind or the relay row it may have made ackable. The rollback story is
/// "the next envelope runs the other engine", nothing finer. Mirrors the shape
/// of `RelayPassEngine`.
enum InboundPathEngine {
    /// The engine that has been receiving mail since before this refactor:
    /// `MeshController.processInboundEnvelope`'s own gate, open, deliver,
    /// carry and re-flood.
    case legacy
    /// `MessageStore.processInboundFrame` — the one transactional inbound
    /// disposition in core — driven from `MeshController` through
    /// `InboundAdapter`.
    case core
}

/// The internal receive settings that are not a person's to change: which
/// engine dispositions an arriving envelope.
///
/// Kept in the same defaults the rest of this device's internal switches live
/// in (`AppDefaults.current`) rather than in the message store, for the reason
/// `RelayEngineSettings` spells out: the whole selection has to be removable
/// once the legacy path is deleted, and a flag that had earned a column in the
/// store's forward-only schema would not be. A preference key simply stops
/// being read.
///
/// The switch is reachable from the internal tools screen, behind the same door
/// as the relay engine switch, because a closed-test build is release-signed
/// and a migration whose flag could only be set by a unit test can never
/// produce the field evidence it exists to produce.
enum InboundEngineSettings {

    /// Absent means `.legacy`.
    private static let pathEngineKey = "cruisemesh.mesh.inboundPathEngineCore"

    /// The engine the *next* arriving envelope will use.
    ///
    /// Read once at the top of `processInboundEnvelope` and not consulted again
    /// while that envelope is being handled, so flipping the switch cannot mix
    /// engines within one frame.
    ///
    /// Defaults to legacy, and stays there until field evidence says otherwise.
    /// A default that quietly moved with a release would make "roll back" mean
    /// "ship again".
    static func pathEngine() -> InboundPathEngine {
        AppDefaults.current.bool(forKey: pathEngineKey) ? .core : .legacy
    }

    static func setPathEngine(_ engine: InboundPathEngine) {
        AppDefaults.current.set(engine == .core, forKey: pathEngineKey)
    }
}

File: android/app/src/main/kotlin/com/cruisemesh/app/mesh/InboundEngineSettings.kt
package com.cruisemesh.app.mesh

import android.content.Context

/**
 * Which engine dispositions one inbound envelope.
 *
 * One selection per envelope, and there is deliberately no third value. A
 * per-stage mix -- core decides dedupe, the shell decides carry -- is the shape
 * of rollback nobody can reason about: the seen-set record, the carry row and
 * the relay ack all have to come from one owner or the DTN ordering rules stop
 * meaning anything. The rollback story is "the next envelope runs the other
 * engine", nothing finer.
 */
enum class InboundEngine {
    /** The engine that has been receiving mail since before this refactor. */
    LEGACY,

    /** `MessageStore.processInboundFrame`, driven through [CoreInboundAdapter]. */
    CORE,
}

/**
 * The internal receive settings that are not a person's to change: which engine
 * dispositions an inbound envelope.
 *
 * Kept in preferences rather than in the message store, for the same reason the
 * relay engine flag is: the whole selection has to be removable once the legacy
 * path is deleted, and a flag that had earned a column in the store's schema
 * would not be. A preference key simply stops being read.
 *
 * The switch is reachable from the internal tools screen. It has to be: a
 * closed-test build is release-signed, `run-as` is refused on it, and a rollout
 * switch that can only be set by a JVM test can never produce the field
 * evidence it exists to produce.
 *
 * There is deliberately no shadow comparison here, and that is the one place
 * this differs from the relay waves. A relay shadow works because the legacy
 * pass leaves a transcript a pure comparator can read after the fact. An
 * inbound envelope has no such transcript: running both engines over one frame
 * would mean opening, carrying and recording it twice, so the comparison would
 * itself be the behaviour change it exists to measure. Evidence comes from
 * running the core engine on a device and watching delivery, not from a
 * side-by-side.
 */
object InboundEngineSettings {

    private const val PREFS_NAME = "cruisemesh_mesh"

    /** Absent means [InboundEngine.LEGACY]. */
    private const val PREF_INBOUND_ENGINE = "inbound_engine_core"

    /**
     * The engine the *next* envelope will use.
     *
     * Read once per envelope, before the disposition starts, and not consulted
     * again -- so flipping the switch changes nothing about an envelope already
     * in flight. Preferences are an in-memory map after the first read, so this
     * is a lock and a lookup on the receive path, not a disk touch.
     *
     * Defaults to legacy, and stays there until field evidence says otherwise.
     * A default that quietly moved with a release would make "roll back" mean
     * "ship again".
     */
    fun inboundEngine(context: Context): InboundEngine =
        if (prefs(context).getBoolean(PREF_INBOUND_ENGINE, false)) {
            InboundEngine.CORE
        } else {
            InboundEngine.LEGACY
        }

    fun setInboundEngine(context: Context, engine: InboundEngine) {
        prefs(context).edit()
            .putBoolean(PREF_INBOUND_ENGINE, engine == InboundEngine.CORE)
            .apply()
    }

    private fun prefs(context: Context) =
        context.applicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
}
